### ISSUE TITLE
Should builds be reproducible?

### ISSUE THREAD

**Issue by dprotaso** [linuxkit/linuxkit#2813]
**Description**

When building ISOs using the `linuxkit build` command multiple builds with the same configuration will result in different artifacts.

**Steps to reproduce the issue:**

In the examples folder
```
$ linuxkit build -format iso-efi sshd.yml
$ mv sshd-efi.iso sshd-efi-first.iso
$ linuxkit build -format iso-efi sshd.yml
$ diffoscope sshd-efi.iso sshd-efi-first.iso
...
```

**Describe the results you received:**

Within the ISO the OCI config.json output will vary across builds due to ordering of nested array elements.

**Describe the results you expected:**

The ISOS to be identical

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by xtuc** [comment]
Linuxkit is meant for "Immutable infrastructure", builds must be deterministic.

The ordering change is probably due to Golang (iteration over a list is guaranty to be random). I think it's fine since the resulting container is still the same?

**Comment by ijc** [comment]
Builds are currently not expected to be bit for bit identical, although they _should_ be functionally identical (modulo things like `files` entries which reference external files).

Personally I think it would be good to strive for bit for bit reproducibility not just in the image build but in package build too (the former is likely to be much easier to achieve), but so fart as I know nobody is working on either right now.

So depending on how you meant "Should" the answer is either "yes" or "no"  ;-)

> The ordering change is probably due to Golang (iteration over a list is guaranty to be random). 

Hopefully you meant "map" and not "list" (or array or slice)?

Can we fix that by sorting some elements or is this something embedded in `encoding/json` and so inaccessible to us?

**Comment by xtuc** [comment]
@ijc yes I meant maps. As far as I know the `encoding/json` serialization also uses Golang's map iteration.

**Comment by deitch** [comment]
They are functionally identical, but it would be good to get it bit-for-bit identical such that `diff` shows identical files.

@ijc do we have any metadata on/in the image that might include a timestamp? Actually, I am sure we do. Some of the image builds for filesystems or partitioning will use date/time stamps as defaults for various in-filesystem bits. I do not recall which off-hand, but I do recall it is there.

If/when we get the tar or kernel+initrd output to be bit-for-bit identical, someone can open an issue to make the built image files identical and assign it to me, and I will track it down. 

Hopefully, but that point, the native build-image-in-go will be complete and merged, and will make it much easier.

**Comment by justincormack** [comment]
It might help if we set the timestamps on the files in the tarballs to a fixed timestamp if they are not files that come from a Docker image, not sure how these are set but may well be wrong. First step is to fix reproducibility for the `linuxkit build -output tar` tarballs, once these are right as that is the base that is used, then work on other output formats.

**Comment by ijc** [comment]
Yes, fixing the timestamps would be a good start. C can't think of an immediately obviously correct date/time to use. e.g. Debian uses the date of the latest changelog entry but we don't have anything similar.

Perhaps step 1 would be to use the mtime of the yml file, which is very imperfect but at least allows interested parties to use `touch(1)` to set it back to some value to repro a build.
